### PR TITLE
Add `ToConstraintFieldGadget`

### PR DESCRIPTION
**Pull request by weikengchen**
This PR adds `ToConstraintFieldGadget` to the refactored Zexe.

### CODE REVIEW

**Comment by weikengchen** [comment]
CI passes.

File: r1cs-std/src/bits/boolean.rs [hunk 1]
```diff
@@ -597,6 +597,14 @@ impl<F: Field> ToBytesGadget<F> for Boolean<F> {
     }
 }
 
+impl<F: PrimeField> ToConstraintFieldGadget<F> for Boolean<F> {
+    #[tracing::instrument(target = "r1cs")]
+    fn to_constraint_field(&self) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        let var = <FpVar<F> as From<Boolean<F>>>::from(self.clone());
```
Comment (Pratyush) [inline r1cs-std/src/bits/boolean.rs]:
```suggestion
        let var = From::from(self.clone());
```

File: r1cs-std/src/fields/cubic_extension.rs [hunk 2]
```diff
@@ -440,6 +441,25 @@ where
...
+impl<BF, P> ToConstraintFieldGadget<P::BasePrimeField> for CubicExtVar<BF, P>
+where
+    BF: FieldVar<P::BaseField, P::BasePrimeField>,
+    for<'a> &'a BF: FieldOpsBounds<'a, P::BaseField, BF>,
+    P: CubicExtVarParams<BF>,
+    BF: ToConstraintFieldGadget<P::BasePrimeField>,
+{
+    #[tracing::instrument(target = "r1cs")]
+    fn to_constraint_field(&self) -> Result<Vec<FpVar<P::BasePrimeField>>, SynthesisError> {
+        let mut res = Vec::<FpVar<P::BasePrimeField>>::new();
```
Comment (Pratyush) [inline r1cs-std/src/fields/cubic_extension.rs]:
```suggestion
        let mut res = Vec::new();
```

File: r1cs-std/src/fields/fp/mod.rs [hunk 3]
```diff
@@ -432,6 +432,14 @@ impl<F: PrimeField> ToBytesGadget<F> for AllocatedFp<F> {
     }
 }
 
+impl<F: PrimeField> ToConstraintFieldGadget<F> for AllocatedFp<F> {
+    #[tracing::instrument(target = "r1cs")]
+    fn to_constraint_field(&self) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        let var = FpVar::from(self.clone());
+        Ok(vec![var])
```
Comment (Pratyush) [inline r1cs-std/src/fields/fp/mod.rs]:
```suggestion
        Ok(vec![self.clone().into()])
```

File: r1cs-std/src/fields/quadratic_extension.rs [hunk 4]
```diff
@@ -430,6 +431,24 @@ where
...
+impl<BF, P> ToConstraintFieldGadget<P::BasePrimeField> for QuadExtVar<BF, P>
+where
+    BF: FieldVar<P::BaseField, P::BasePrimeField>,
+    for<'a> &'a BF: FieldOpsBounds<'a, P::BaseField, BF>,
+    P: QuadExtVarParams<BF>,
+    BF: ToConstraintFieldGadget<P::BasePrimeField>,
+{
+    #[tracing::instrument(target = "r1cs")]
+    fn to_constraint_field(&self) -> Result<Vec<FpVar<P::BasePrimeField>>, SynthesisError> {
+        let mut res = Vec::<FpVar<P::BasePrimeField>>::new();
```
Comment (Pratyush) [inline r1cs-std/src/fields/quadratic_extension.rs]:
```suggestion
        let mut res = Vec::new();
```

File: r1cs-std/src/groups/curves/twisted_edwards/mod.rs [hunk 5]
```diff
@@ -646,6 +647,25 @@ where
...
+where
+    P: TEModelParameters,
+    F: FieldVar<P::BaseField, <P::BaseField as Field>::BasePrimeField>,
+    for<'a> &'a F: FieldOpsBounds<'a, P::BaseField, F>,
+    F: ToConstraintFieldGadget<<P::BaseField as Field>::BasePrimeField>,
+{
+    fn to_constraint_field(
+        &self,
+    ) -> Result<Vec<FpVar<<P::BaseField as Field>::BasePrimeField>>, SynthesisError> {
+        let mut res = Vec::<FpVar<<P::BaseField as Field>::BasePrimeField>>::new();
```
Comment (Pratyush) [inline r1cs-std/src/groups/curves/twisted_edwards/mod.rs]:
```suggestion
        let mut res = Vec::new();
```

**Comment by weikengchen** [comment]
CI passes!

**Pull request merged**